src/tapedfunction.jl

```diff
@@ -168,127 +154,180 @@ function (instr::Instruction{F})() where F
     return (args...) -> begin
-        @assert length(args) + 1 == length(ins.args)
-        ins.args[1].val = ins.tape.func
+        funcbox.val = taped.func
+        @assert length(args) == taped.arity
         for i in 1:length(args)
-            ins.args[i + 1].val = val(args[i]) # fill the boxes
+            argsbox[i].val = val(args[i]) # fill the boxes
         end
     end
 end
 
-function translate!(taped::Taped, ir::IRTools.IR)
+kind(e) = Val(:nonexpr)
+kind(e::Expr) = Val(e.head)
+
+function translate!(taped::Taped, ir::Core.CodeInfo)
     tape = taped.tape
-    boxes = Dict{IRTools.Variable, Box{Any}}()
+    boxes = Dict{Symbol, Box{<:Any}}()
     _box = (x) -> arg_boxer(x, boxes)
-    for (blk_id, blk) in enumerate(IRTools.blocks(ir))
-        # blocks
-        blk_args = IRTools.arguments(blk)
-        push!(tape, BlockInstruction(blk_id, map(_box, blk_args), taped))
-        # `+ 1` because we will have an extra ins at the beginning
-        taped.block_map[blk_id] = length(tape) + 1
-        # normal instructions
-        for (x, st) in blk
-            if Meta.isexpr(st.expr, :call)
-                args = map(_box, st.expr.args)
-                # args[1] is the function
-                f = args[1]
-                ins = Instruction(f, args[2:end] |> Tuple,
-                                  _box(x), taped)
-                push!(tape, ins)
-            elseif Meta.isexpr(st.expr, :new)
-                args = map(_box, st.expr.args)
-                ins = Instruction(__new__, args |> Tuple, _box(x), taped)
-                push!(tape, ins)
-            elseif isa(st.expr, GlobalRef)
-                ins = Instruction(val, (st.expr,), _box(x), taped)
-                push!(tape, ins)
-            else
-                @error "Unknown IR code: " st
-            end
-        end
 
-        # branches (including `return`)
-        for br in IRTools.branches(blk)
-            if br.condition === nothing && br.block == 0 # a return
-                ins = ReturnInstruction(_box(br.args[1]), taped)
-                push!(tape, ins)
-            else
-                cond = br.condition === nothing ? false : _box(br.condition)
-                isa(cond, Bool) && (cond = Box{Any}(cond)) # unify the condiftion type
-                ins = BranchInstruction(
-                    cond, br.block, map(_box, br.args), taped)
-                push!(tape, ins)
-            end
-        end
+    builder = InstructionBuilder(taped, ir, _box)
+
+    for (idx, line) in enumerate(ir.code)
+        isa(line, Core.Const) && (line = line.val) # unbox Core.Const
+        builder(Core.SSAValue(idx), line, kind(line))
     end
-    init_ins = Instruction(args_initializer(tape[1]), tuple(tape[1].args[2:end]...),
-                           Box{Any}(nothing), taped)
+
+    init_ins = Instruction(
+        args_initializer(taped, boxes),
+        tuple((Box{Any}(nothing) for _ in 1:taped.arity)...),
```

Comment:
Can't be inferred either (since arity is not known at compile time) but avoids splatting:
```suggestion
        ntuple(_ -> Box{Any}(nothing), taped.arity),
```